src/System.Private.ServiceModel/src/System/ServiceModel/Channels/CommunicationObject.cs

```diff
@@ -769,15 +773,23 @@ internal protected virtual Task OnCloseAsync(TimeSpan timeout)
 #endif
 
             // If we have already executed this code path, just return a completed Task.
-            // It means this type did not override OnCloseAsync or OnBeginClose.
+            // This is a safeguard in case custom CommunicationObjects attempt to do this multiple times.
             if (_onCloseAsyncCalled)
             {
                 return TaskHelpers.CompletedTask();
             }
 
-            // External types cannot override this method because it is not public.
-            // Redirect this call to the APM path.
             _onCloseAsyncCalled = true;
+
+            // External types cannot override this method because it is not public.
+            // If the caller started this open synchronously, we must use the synchronous
+            // methods they may have overridden.
+            if (_isSynchronousClose)
+            {
+                return Task.Run(() => OnClose(timeout));
```

Comment:
This will cause a lot more threads to be used. In a channel factory with 3 channel layers, each layer will cause an extra thread to be used. This will cause thread pool starvation and complete process stalling as we wait for the threadpool to add more threads.